packages/next/build/swc/index.js

```diff
@@ -1,46 +1,81 @@
 import { patchIncorrectLockfile } from '../../lib/patch-incorrect-lockfile'
+import { downloadWasmSwc } from '../../lib/download-wasm-swc'
+import { version as nextVersion } from 'next/package.json'
 
 const ArchName = arch()
 const PlatformName = platform()
 const triples = platformArchTriples[PlatformName][ArchName] || []
 
 let nativeBindings
 let wasmBindings
+let downloadWasmPromise
+let pendingBindings
 export const lockfilePatchPromise = {}
 
 async function loadBindings() {
-  if (!lockfilePatchPromise.cur) {
-    // always run lockfile check once so that it gets patched
-    // even if it doesn't fail to load locally
-    lockfilePatchPromise.cur = patchIncorrectLockfile(process.cwd()).catch(
-      console.error
-    )
+  if (pendingBindings) {
+    return pendingBindings
   }
+  pendingBindings = new Promise(async (resolve, reject) => {
+    if (!lockfilePatchPromise.cur) {
+      // always run lockfile check once so that it gets patched
+      // even if it doesn't fail to load locally
+      lockfilePatchPromise.cur = patchIncorrectLockfile(process.cwd()).catch(
+        console.error
+      )
+    }
 
-  let attempts = []
-  try {
-    return loadNative()
-  } catch (a) {
-    attempts = attempts.concat(a)
-  }
+    let attempts = []
+    try {
+      return resolve(loadNative())
+    } catch (a) {
+      attempts = attempts.concat(a)
+    }
 
-  // TODO: fetch wasm and fallback when loading native fails
-  // so that users aren't blocked on this, we still want to
-  // report the native load failure so we can patch though
-  try {
-    let bindings = await loadWasm()
-    return bindings
-  } catch (a) {
-    attempts = attempts.concat(a)
-  }
+    try {
+      let bindings = await loadWasm()
+      eventSwcLoadFailure({ wasm: 'enabled' })
+      return resolve(bindings)
+    } catch (a) {
+      attempts = attempts.concat(a)
+    }
 
-  logLoadFailure(attempts)
+    try {
+      // if not installed already download wasm package on-demand
+      // we download to a custom directory instead of to node_modules
+      // as node_module import attempts are cached and can't be re-attempted
+      // x-ref: https://github.com/nodejs/modules/issues/307
+      const wasmDirectory = path.join(
+        path.dirname(require.resolve('next/package.json')),
+        'wasm'
+      )
+      if (!downloadWasmPromise) {
+        downloadWasmPromise = downloadWasmSwc(nextVersion, wasmDirectory)
+      }
+      await downloadWasmPromise
+      let bindings = await loadWasm(pathToFileURL(wasmDirectory).href)
+      eventSwcLoadFailure({ wasm: 'fallback' })
+
+      // still log native load attempts so user is
+      // aware it failed and should be fixed
+      for (const attempt of attempts) {
+        Log.warn(attempt)
+      }
+      return resolve(bindings)
+    } catch (a) {
+      attempts = attempts.concat(a)
+    }
+
+    logLoadFailure(attempts, true)
+  })
+  return pendingBindings
 }
 
 function loadBindingsSync() {
```

Comment:
Will this fallback only work for async bindings calls?

Reply:
Currently yes since it requires fetching the wasm package so we aren't able to fallback for the `jest` case unless we load the bindings eagerly similar to what @huozhi is describing below.